internal/handlers/common/projection.go

```diff
@@ -133,102 +171,209 @@ func projectDocument(inclusion bool, doc *types.Document, projection *types.Docu
+	found = -1 // >= 0 means found
 
-		case *types.Array:
-			panic("unexpected code")
+	// for sure it's here - see code above
+	docValueA := must.NotFail(doc.GetByPath(k1))
 
-		default: // field2: value
-			found = -1 // >= 0 means found
+	// $elemMatch works only for arrays, it must be an array
+	docValueArray, ok := docValueA.(*types.Array)
+	if !ok {
+		doc.Remove(k1)
+		return
+	}
 
-			for j := 0; j < docValueArray.Len(); j++ {
-				var cmpVal any
-				cmpVal, err = docValueArray.Get(j)
-				if err != nil {
-					continue
-				}
-				switch cmpVal := cmpVal.(type) {
-				case *types.Document:
-					docVal, err := cmpVal.Get(k2ConditionField)
-					if err != nil {
-						doc.RemoveByPath(k1, strconv.Itoa(j))
-						continue
+	for k2, condition := range conditions.Map() {
+		switch condition := condition.(type) {
+		// in condition: { $eq: 42 }
+		case *types.Document:
+			for operand, value := range condition.Map() {
+				switch operand {
+				case "$eq":
+					found, err = findInArray(k1, k2, value, doc, []types.CompareResult{types.Equal})
+
+				case "$ne":
+					found, err = findInArray(k1, k2, value, doc, []types.CompareResult{types.Less, types.Greater})
+
+				case "$gt":
+					found, err = findInArray(k1, k2, value, doc, []types.CompareResult{types.Greater})
+
+				case "$gte":
+					found, err = findInArray(k1, k2, value, doc, []types.CompareResult{types.Greater, types.Equal})
+
+				case "$lt":
+					found, err = findInArray(k1, k2, value, doc, []types.CompareResult{types.Less})
+
+				case "$lte":
+					found, err = findInArray(k1, k2, value, doc, []types.CompareResult{types.Less, types.Equal})
+
+				case "$nin":
+					switch inValue := value.(type) {
+					case *types.Array:
+						for i := 0; i < inValue.Len(); i++ {
+							x := must.NotFail(inValue.Get(i))
+							found, err = findInArray(k1, k2, x, doc,
+								[]types.CompareResult{types.Less, types.Greater, types.NotEqual},
+							)
+							if found >= 0 {
+								return
+							}
+						}
+					default:
+						err = NewErrorMsg(ErrBadValue, "$nin needs an array")
+						return
 					}
-					if types.Compare(docVal, elemMatchFieldCondition) == types.Equal {
-						// elemMatch to return first matching, all others are to be removed
-						found = j
-						break
+					if found < 0 {
+						return
 					}
-					doc.RemoveByPath(k1, strconv.Itoa(j))
-					j = j - 1
-				}
-			}
 
-			if found < 0 {
-				return
+				case "$in":
+					switch inValue := value.(type) {
+					case *types.Array:
+						for i := 0; i < inValue.Len(); i++ {
+							x := must.NotFail(inValue.Get(i))
+							found, err = findInArray(k1, k2, x, doc, []types.CompareResult{types.Equal})
+							if found >= 0 {
+								return
+							}
+						}
+					default:
+						err = NewErrorMsg(ErrBadValue, "array values supported for $in only")
+						return
+					}
+					if found < 0 {
+						return
+					}
+
+					// operand is not an operand possible: <scalar value> OR field: {nested projection}
```

Comment:
```suggestion
				// operand is not possible: <scalar value> OR field: {nested projection}
```

Reply:
no, please read the comment. 
the variable name is operand, but here the condition detects that it's not an operand.